types/wlr_surface_layers.c

```diff
@@ -0,0 +1,287 @@
+	} else {
+		*y = (double)(output_height - height) / 2;
+	}
+}
+
+void wlr_layer_surface_configure(struct wlr_layer_surface *layer_surface,
+		int32_t width, int32_t height) {
+	layer_surface_send_configure(layer_surface->resource, width, height);
+}
+
+static void layer_surface_destroy(struct wlr_layer_surface *surface) {
+	wl_signal_emit(&surface->events.destroy, surface);
+	wl_list_remove(&surface->link);
+	wl_resource_set_user_data(surface->resource, NULL);
+	free(surface->current);
+	free(surface->pending);
+	free(surface);
+}
+
+static void layer_surface_resource_destroy(struct wl_resource *resource) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	if (surface != NULL) {
+		layer_surface_destroy(surface);
+	}
+}
+
+static const struct layer_surface_interface layer_surface_impl = {
+	.set_interactivity = layer_surface_set_interactivity,
+	.set_anchor = layer_surface_set_anchor,
+	.set_exclusive_zone = layer_surface_set_exclusive_zone,
+	.set_margin = layer_surface_set_margin,
+};
+
+static void handle_layer_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_layer_surface *surface =
+		wl_container_of(listener, surface, surface_commit_listener);
+	wlr_layer_surface_move_state(surface->pending, surface->current);
+	wl_signal_emit(&surface->events.commit, surface);
+}
+
+static void handle_layer_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_layer_surface *surface =
+		wl_container_of(listener, surface, surface_destroy_listener);
+	layer_surface_destroy(surface);
+}
+
+static void surface_layers_get_layer_surface(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id,
+		struct wl_resource *surface_resource,
+		struct wl_resource *output_resource, uint32_t layer) {
+	struct wlr_surface_layers *surface_layers =
+		wl_resource_get_user_data(resource);
+	struct wlr_surface *surface = wl_resource_get_user_data(surface_resource);
+	struct wlr_output *output = wl_resource_get_user_data(output_resource);
+
+	if (layer > SURFACE_LAYERS_LAYER_OVERLAY) {
+		wl_resource_post_error(resource, SURFACE_LAYERS_ERROR_INVALID_LAYER,
+			"Unknown layer");
+		return;
+	}
+	if (wlr_surface_set_role(surface, surface_layers_role, resource,
+			SURFACE_LAYERS_ERROR_ROLE)) {
+		return;
+	}
+
+	struct wlr_layer_surface *layer_surface =
+		calloc(1, sizeof(struct wlr_layer_surface));
+	if (layer_surface == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	layer_surface->resource = wl_resource_create(client,
+		&layer_surface_interface, wl_resource_get_version(resource), id);
+	wlr_log(L_DEBUG, "new layer_surface %p (res %p)", layer_surface,
+		layer_surface->resource);
+	wl_resource_set_implementation(layer_surface->resource, &layer_surface_impl,
+		layer_surface, layer_surface_resource_destroy);
+
+	layer_surface->surface_layers = surface_layers;
+	layer_surface->surface = surface;
+	layer_surface->output = output;
```

Comment:
This will become invalid when the output is hotplugged, which will lead to memory errors.